### DISCUSSION TITLE
Docker image for ARM?

### DISCUSSION THREAD

**Discussion by manojchandrashekar** [outline/outline#1936] [Ideas]
I was trying to setup outline using docker on a raspberry pi and it fails with:

`standard_init_linux.go:219: exec user process caused: exec format error`

As far as I know, this is because of the mismatch between docker base image architecture and host architecture.

Would be great to have an official docker image for ARM.

**Comment by astwyg** [comment]
same problem, I want to run outline on a Arm server.

any guide to install without docker?

**Reply by tommoor** [reply]
You're probably the first to try, let us know how you get on 😄 

I don't see any reason there would be a problem. I develop locally on an M1 machine without issues.

**Reply by manojchandrashekar** [reply]
I was able to get it running on a raspberry pi 4. But I am using docker. I will try to document what I did later today.

Shouldn’t be very difficult to run without docker, but it’s definitely easier with it.

**Reply by a7bf9** [reply]
Can you give any updates on how you managed to get it to work @manojchandrashekar? A couple months later and I am at the same problem but can't seem to figure it out sadly.

**Reply by manojchandrashekar** [reply]
Please check my fork here: https://github.com/manojchandrashekar/outline (start here: https://github.com/manojchandrashekar/outline/blob/8bc44a5290e8aeb64207a80caf4871d7bad36e1d/docker-compose.yml)

Note: The code is several commits behind master.

**Reply by a7bf9** [reply]
Thank you!

**Reply by tommoor** [reply]
FWIW I don't see anything ARM specific about these changes, it's just a modified version of the development docker-compose file that's a little more suitable for production 🤷

**Reply by manojchandrashekar** [reply]
@tommoor Have you checked the Dockerfile? 

Yes, I have modified it to suit my deployment needs. Feel free to fork and make your own changes.

**Reply by tommoor** [reply]
Ah I see, I wonder if Dockerhub supports building ARM images – it would be reasonable to have an arm dockerfile in the project and have it autobuild along with the regular image if that's all it takes

**Reply by manojchandrashekar** [reply]
Oops, sorry @tommoor - I replied on the mobile and didn't realize that you were a maintainer. :)

Yes, I suppose the Dockerfile needs to be made "intelligent" to use the right base image depending on the host architecture. I remember looking into it a while back.

This is what I had in my bookmarks, hope it helps: 

* https://stackoverflow.com/questions/53048942/is-it-possible-to-get-the-architecture-of-the-docker-engine-in-a-dockerfile
* https://docs.docker.com/desktop/multi-arch/

**Reply by jpvalery** [reply]
@manojchandrashekar how's your experience running Outline on a Pi? I'd be interested in running it on my Pi 3 B+ but I'm a bit confused as to what should I do to get it running. Any help or pointers will be appreciated 🙏

**Reply by manojchandrashekar** [reply]
@jpvalery I am running it on Pi 4. And it works pretty well. One option is to install docker and docker-compose on the Pi. Use the modified docker files (https://github.com/manojchandrashekar/outline/blob/8bc44a5290e8aeb64207a80caf4871d7bad36e1d/docker-compose.yml), build and run the image there. 

Note: The Dockerfile isn't much different, it just uses ARM versions of the base images and it was created several months ago. So, you might want to create this setup on the latest code from outline, so you take advantage of the latest changes.

**Reply by jpvalery** [reply]
@manojchandrashekar Thanks, really appreciate the pointers.

I'm still facing different errors and if anyone wants to check my progress:  https://github.com/jpvalery/outline/issues/1 

I needed to upgrade to a Pi 4 2Gb to get the compose to complete and now trying to remove error after error.

**Reply by jpvalery** [reply]
@tommoor I'm running into an issue now on the `server/index.js` as you can see below—if you have any suggestions or ideas, I'd take them 😇 

```
outline_1   | yarn run v1.22.5
outline_1   | $ node ./build/server/index.js
outline_1   | internal/validators.js:124
outline_1   |     throw new ERR_INVALID_ARG_TYPE(name, 'string', value);
outline_1   |     ^
outline_1   | 
outline_1   | TypeError [ERR_INVALID_ARG_TYPE]: The "url" argument must be of type string. Received undefined
outline_1   |     at validateString (internal/validators.js:124:11)
outline_1   |     at Url.parse (url.js:170:3)
outline_1   |     at Object.urlParse [as parse] (url.js:157:13)
outline_1   |     at new Sequelize (/opt/outline/node_modules/sequelize/lib/sequelize.js:185:28)
outline_1   |     at Object.<anonymous> (/opt/outline/build/server/sequelize.js:26:19)
outline_1   |     at Module._compile (internal/modules/cjs/loader.js:1072:14)
outline_1   |     at Object.Module._extensions..js (internal/modules/cjs/loader.js:1101:10)
outline_1   |     at Module.load (internal/modules/cjs/loader.js:937:32)
outline_1   |     at Function.Module._load (internal/modules/cjs/loader.js:778:12)
outline_1   |     at Module.require (internal/modules/cjs/loader.js:961:19)
outline_1   |     at require (internal/modules/cjs/helpers.js:92:18)
outline_1   |     at Object.<anonymous> (/opt/outline/build/server/models/ApiKey.js:10:18)
outline_1   |     at Module._compile (internal/modules/cjs/loader.js:1072:14)
outline_1   |     at Object.Module._extensions..js (internal/modules/cjs/loader.js:1101:10)
outline_1   |     at Module.load (internal/modules/cjs/loader.js:937:32)
outline_1   |     at Function.Module._load (internal/modules/cjs/loader.js:778:12) {
outline_1   |   code: 'ERR_INVALID_ARG_TYPE'
outline_1   | }
outline_1   | error Command failed with exit code 1.
outline_1   | info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
outline_outline_1 exited with code 1
```
